controllers/buildpipeline/buildpipeline_adapter.go

```diff
@@ -108,6 +108,36 @@ func (a *Adapter) EnsureSnapshotExists() (result controller.OperationResult, err
 		return controller.ContinueProcessing()
 	}
 
+	if _, found := a.pipelineRun.ObjectMeta.Annotations[tekton.SnapshotNameLabel]; found {
+		a.logger.Info("The build pipelineRun is already associated with existing Snapshot via annotation",
+			"snapshot.Name", a.pipelineRun.ObjectMeta.Annotations[tekton.SnapshotNameLabel])
+		canRemoveFinalizer = true
+		return controller.ContinueProcessing()
+	}
+
+	existingSnapshots, err := a.loader.GetAllSnapshotsForBuildPipelineRun(a.client, a.context, a.pipelineRun)
+	if err != nil {
+		a.logger.Error(err, "Failed to fetch Snapshots for the build pipelineRun")
+		return controller.RequeueWithError(err)
+	}
+	if len(*existingSnapshots) > 0 {
+		if len(*existingSnapshots) == 1 {
+			existingSnapshot := (*existingSnapshots)[0]
+			a.logger.Info("There is an existing Snapshot associated with this build pipelineRun, but the pipelineRun is not yet annotated",
+				"snapshot.Name", existingSnapshot.Name)
+			err := a.annotateBuildPipelineRunWithSnapshot(a.pipelineRun, &existingSnapshot)
+			if err != nil {
+				a.logger.Error(err, "Failed to update the build pipelineRun with snapshot name",
+					"pipelineRun.Name", a.pipelineRun.Name)
+				return controller.RequeueWithError(err)
+			}
+		} else {
+			a.logger.Info("The build pipelineRun is already associated with more than one existing Snapshot")
```

Comment:
When would we come inside this else condition? (or is it possible to have more than 1 Snapshots for a build PLR based on the current logic?)

Reply:
It shouldn't be possible aside for possible edge cases caused by extreme API latency. I thought it would be good to cover (and log) that remote possibility just in case.